File: packages/cli/src/services/http-client.ts
import { WebSocket } from "ws";
import { attachLocalHttpForwarder } from "#app/client/local-http-forwarder.ts";
import { attachLocalWebSocketForwarder } from "#app/client/local-websocket-forwarder.ts";
import { ProxerError } from "#app/lib/error.ts";
import {
  type LogRoute,
  type RuntimeLogger,
  sanitizeLogUrl,
} from "#app/lib/logging.ts";
import { createWebSocketTunnelConnection } from "#app/protocol/tunnel-connection.ts";

export type HttpClientRouteRequest =
  | { readonly type: "auto" }
  | { readonly type: "root" }
  | { readonly type: "subdomain"; readonly subdomain: string };

export type HttpClientConfig = {
  readonly basicAuth?: {
    readonly password: string;
    readonly username?: string;
  };
  readonly localPort: number;
  readonly serverUrl: string;
  readonly route?: HttpClientRouteRequest;
  readonly token?: string;
  readonly heartbeatIntervalMs?: number;
  readonly heartbeatTimeoutMs?: number;
  readonly logger?: RuntimeLogger;
  readonly reconnectDelayMs?: number;
};

type HttpClientBasicAuth = NonNullable<HttpClientConfig["basicAuth"]>;

type RegisteredRoute =
  | { readonly type: "root" }
  | { readonly type: "subdomain"; readonly subdomain: string };

export type RunningTunnelClient = {
  readonly subdomain?: string;
  close(): Promise<void>;
};

const openWebSocket = async (url: string): Promise<WebSocket> => {
  const socket = new WebSocket(url);

  await new Promise<void>((resolve, reject) => {
    socket.once("open", resolve);
    socket.once("error", reject);
  });

  return socket;
};

const DEFAULT_HEARTBEAT_INTERVAL_MS = 30_000;
const DEFAULT_HEARTBEAT_TIMEOUT_MS = 10_000;
const DEFAULT_RECONNECT_DELAY_MS = 1_000;
const silentLogger: RuntimeLogger = {
  info() {},
  error() {},
};

const formatRouteName = (
  route: HttpClientRouteRequest | RegisteredRoute,
): string => {
  if (route.type === "subdomain") {
    return route.subdomain;
  }

  return route.type;
};

const buildRegisterRouteFields = (
  route: HttpClientRouteRequest,
): { readonly root?: true; readonly subdomain?: string } => {
  if (route.type === "root") {
    return { root: true };
  }

  if (route.type === "subdomain") {
    return { subdomain: route.subdomain };
  }

  return {};
};

const resolveRegisteredRoute = (
  frameSubdomain: string | undefined,
  requestedRoute: HttpClientRouteRequest,
): RegisteredRoute => {
  if (requestedRoute.type === "auto") {
    if (frameSubdomain) {
      return { type: "subdomain", subdomain: frameSubdomain };
    }

    throw new ProxerError('Registered unexpected tunnel "root"');
  }

  if (requestedRoute.type === "root") {
    if (frameSubdomain === undefined) {
      return { type: "root" };
    }

    throw new ProxerError(`Registered unexpected tunnel "${frameSubdomain}"`);
  }

  if (frameSubdomain === requestedRoute.subdomain) {
    return { type: "subdomain", subdomain: frameSubdomain };
  }

  throw new ProxerError(
    `Registered unexpected tunnel "${frameSubdomain ?? "root"}"`,
  );
};

const registerConnection = async ({
  basicAuth,
  connection,
  route,
  socket,
  token,
}: {
  readonly basicAuth?: HttpClientBasicAuth;
  readonly connection: ReturnType<typeof createWebSocketTunnelConnection>;
  readonly route: HttpClientRouteRequest;
  readonly socket: WebSocket;
  readonly token?: string;
}): Promise<RegisteredRoute> => {
  let removeRegistrationFrameListener: () => void = () => {};
  let removeRegistrationCloseListener: () => void = () => {};

  try {
    return await new Promise<RegisteredRoute>((resolve, reject) => {
      removeRegistrationFrameListener = connection.onFrame((frame) => {
        if (frame.type === "registered") {
          try {
            resolve(resolveRegisteredRoute(frame.subdomain, route));
          } catch (error) {
            reject(error instanceof Error ? error : new Error(String(error)));
          }
          return;
        }

        if (frame.type === "error" && frame.streamId === "registration") {
          reject(new ProxerError(frame.message));
        }
      });
      removeRegistrationCloseListener = connection.onClose((error) => {
        reject(error ?? new ProxerError("Tunnel registration closed"));
      });
      void connection
        .send({
          type: "register",
          ...(basicAuth ? { basicAuth } : {}),
          ...buildRegisterRouteFields(route),
          token,
        })
        .catch((error: unknown) => {
          reject(error instanceof Error ? error : new Error(String(error)));
        });
    });
  } catch (error) {
    socket.close();
    throw error;
  } finally {
    removeRegistrationFrameListener();
    removeRegistrationCloseListener();
  }
};

const startHeartbeat = ({
  intervalMs,
  socket,
  timeoutMs,
}: {
  readonly intervalMs: number;
  readonly socket: WebSocket;
  readonly timeoutMs: number;
}): (() => void) => {
  if (intervalMs <= 0) {
    return () => {};
  }

  let heartbeatTimeout: ReturnType<typeof setTimeout> | undefined;
  const clearHeartbeatTimeout = (): void => {
    if (heartbeatTimeout === undefined) {
      return;
    }

    clearTimeout(heartbeatTimeout);
    heartbeatTimeout = undefined;
  };
  const onPong = (): void => {
    clearHeartbeatTimeout();
  };
  const interval = setInterval(() => {
    if (socket.readyState !== WebSocket.OPEN) {
      return;
    }

    if (heartbeatTimeout !== undefined) {
      socket.terminate();
      return;
    }

    socket.ping();
    heartbeatTimeout = setTimeout(() => {
      heartbeatTimeout = undefined;
      socket.terminate();
    }, timeoutMs);
  }, intervalMs);

  socket.on("pong", onPong);

  return () => {
    clearInterval(interval);
    clearHeartbeatTimeout();
    socket.off("pong", onPong);
  };
};

type ActiveTunnelConnection = {
  readonly connection: ReturnType<typeof createWebSocketTunnelConnection>;
  readonly detachLocalHttpForwarder: () => void;
  readonly detachLocalWebSocketForwarder: () => void;
  removeLifecycleCloseListener: () => void;
  readonly route: RegisteredRoute;
  readonly socket: WebSocket;
  readonly stopHeartbeat: () => void;
};

export const startHttpTunnelClient = async ({
  basicAuth,
  heartbeatIntervalMs = DEFAULT_HEARTBEAT_INTERVAL_MS,
  heartbeatTimeoutMs = DEFAULT_HEARTBEAT_TIMEOUT_MS,
  localPort,
  logger = silentLogger,
  reconnectDelayMs = DEFAULT_RECONNECT_DELAY_MS,
  route = { type: "auto" },
  serverUrl,
  token,
}: HttpClientConfig): Promise<RunningTunnelClient> => {
  const tunnelToken = token?.trim();
  if (!tunnelToken) {
    throw new ProxerError("token is required");
  }
  const logServerUrl = sanitizeLogUrl(serverUrl);
  let routeRequest = route;
  let activeRoute: RegisteredRoute | undefined;

  let activeConnection: ActiveTunnelConnection | undefined;
  let closing = false;
  let reconnectTimer: ReturnType<typeof setTimeout> | undefined;

  const cleanupActiveConnection = (): void => {
    if (!activeConnection) {
      return;
    }

    activeConnection.removeLifecycleCloseListener();
    activeConnection.stopHeartbeat();
    activeConnection.detachLocalWebSocketForwarder();
    activeConnection.detachLocalHttpForwarder();
    activeConnection = undefined;
  };

  const connect = async (): Promise<ActiveTunnelConnection> => {
    logger.info(
      `connecting server=${logServerUrl} route=${formatRouteName(routeRequest)}`,
    );
    const socket = await openWebSocket(serverUrl);
    logger.info(`connected server=${logServerUrl}`);
    const connection = createWebSocketTunnelConnection(socket);
    const registeredRoute = await registerConnection({
      basicAuth,
      connection,
      route: routeRequest,
      socket,
      token: tunnelToken,
    });
    activeRoute = registeredRoute;
    if (routeRequest.type === "auto" && registeredRoute.type === "subdomain") {
      routeRequest = {
        type: "subdomain",
        subdomain: registeredRoute.subdomain,
      };
    }
    logger.info(`registered route=${formatRouteName(registeredRoute)}`);
    const logRoute: LogRoute = registeredRoute;
    const detachLocalHttpForwarder = attachLocalHttpForwarder({
      connection,
      localPort,
      logger,
      route: logRoute,
    });
    const detachLocalWebSocketForwarder = attachLocalWebSocketForwarder({
      connection,
      localPort,
      logger,
      route: logRoute,
    });
    const stopHeartbeat = startHeartbeat({
      intervalMs: heartbeatIntervalMs,
      socket,
      timeoutMs: heartbeatTimeoutMs,
    });
    return {
      connection,
      detachLocalHttpForwarder,
      detachLocalWebSocketForwarder,
      removeLifecycleCloseListener: () => {},
      route: registeredRoute,
      socket,
      stopHeartbeat,
    };
  };

  const activateConnection = (active: ActiveTunnelConnection): void => {
    activeConnection = active;
    active.removeLifecycleCloseListener = active.connection.onClose(() => {
      if (activeConnection !== active) {
        return;
      }

      logger.info(`disconnected route=${formatRouteName(active.route)}`);
      cleanupActiveConnection();
      scheduleReconnect();
    });
  };

  const reconnect = async (): Promise<void> => {
    if (closing) {
      return;
    }

    try {
      const nextConnection = await connect();
      activateConnection(nextConnection);
      logger.info(`reconnected route=${formatRouteName(nextConnection.route)}`);
      if (closing) {
        const staleConnection = nextConnection;
        cleanupActiveConnection();
        await staleConnection.connection.close();
      }
    } catch {
      scheduleReconnect();
    }
  };

  function scheduleReconnect(): void {
    if (closing || reconnectTimer !== undefined) {
      return;
    }

    logger.info(`reconnecting in ${reconnectDelayMs}ms`);
    reconnectTimer = setTimeout(() => {
      reconnectTimer = undefined;
      void reconnect();
    }, reconnectDelayMs);
  }

  activateConnection(await connect());

  return {
    subdomain:
      activeRoute?.type === "subdomain" ? activeRoute.subdomain : undefined,
    async close() {
      closing = true;
      if (reconnectTimer !== undefined) {
        clearTimeout(reconnectTimer);
        reconnectTimer = undefined;
      }

      const connectionToClose = activeConnection;
      cleanupActiveConnection();
      if (!connectionToClose) {
        return;
      }

      if (connectionToClose.socket.readyState === WebSocket.CLOSED) {
        return;
      }

      await connectionToClose.connection.close();
    },
  };
};

File: packages/cli/lib/src/services/http_client.dart
import 'dart:async';
import 'dart:io';

import 'package:proxer/src/services/client/local_forwarders.dart';
import 'package:proxer/src/services/protocol/frame.dart';
import 'package:proxer/src/services/protocol/tunnel_connection.dart';
import 'package:proxer/src/util/error.dart';

sealed class HttpClientRouteRequest {
  const HttpClientRouteRequest();
}

class AutoRouteRequest extends HttpClientRouteRequest {
  const AutoRouteRequest();
}

class RootRouteRequest extends HttpClientRouteRequest {
  const RootRouteRequest();
}

class SubdomainRouteRequest extends HttpClientRouteRequest {
  const SubdomainRouteRequest(this.subdomain);

  final String subdomain;
}

class HttpClientConfig {
  const HttpClientConfig({
    required this.localPort,
    required this.serverUrl,
    required this.token,
    this.basicAuth,
    this.mode = RouteMode.single,
    this.route = const AutoRouteRequest(),
    this.heartbeatIntervalMs = 30000,
    this.reconnectDelayMs = 1000,
    this.log,
    this.logError,
  });

  final BasicAuthConfig? basicAuth;
  final int localPort;
  final String serverUrl;
  final RouteMode mode;
  final HttpClientRouteRequest route;
  final String token;
  final int heartbeatIntervalMs;
  final int reconnectDelayMs;
  final void Function(String message)? log;
  final void Function(String message)? logError;
}

class RunningTunnelClient {
  const RunningTunnelClient({required this.subdomain, required this.close});

  final String? subdomain;
  final Future<void> Function() close;
}

class _ActiveConnection {
  const _ActiveConnection({
    required this.connection,
    required this.forwarders,
    required this.route,
  });

  final TunnelConnection connection;
  final LocalForwarders forwarders;
  final HttpClientRouteRequest route;
}

Future<RunningTunnelClient> startHttpTunnelClient(
  HttpClientConfig config,
) async {
  final token = config.token.trim();
  if (token.isEmpty) {
    throw ProxerError('token is required');
  }
  var requestedRoute = config.route;
  _ActiveConnection? active;
  Timer? reconnectTimer;
  var closing = false;

  Future<_ActiveConnection> connect() async {
    config.log?.call('connecting server=${config.serverUrl}');
    final socket = await WebSocket.connect(config.serverUrl);
    if (config.heartbeatIntervalMs > 0) {
      socket.pingInterval = Duration(milliseconds: config.heartbeatIntervalMs);
    }
    final connection = TunnelConnection(socket);
    final completer = Completer<RegisteredFrame>();
    late final StreamSubscription<TunnelFrame> frames;
    late final StreamSubscription<Object?> closed;
    frames = connection.frames.listen((frame) {
      if (frame is RegisteredFrame && !completer.isCompleted) {
        completer.complete(frame);
      } else if (frame is ErrorFrame &&
          frame.streamId == 'registration' &&
          !completer.isCompleted) {
        completer.completeError(ProxerError(frame.message));
      }
    });
    closed = connection.closed.listen((error) {
      if (!completer.isCompleted) {
        completer.completeError(
          error ?? ProxerError('Tunnel registration closed'),
        );
      }
    });
    final routeForRegistration = requestedRoute;
    await connection.send(
      RegisterFrame(
        root: routeForRegistration is RootRouteRequest ? true : null,
        subdomain: routeForRegistration is SubdomainRouteRequest
            ? routeForRegistration.subdomain
            : null,
        mode: config.mode,
        token: token,
        basicAuth: config.basicAuth,
      ),
    );
    RegisteredFrame registered;
    try {
      registered = await completer.future;
    } on Object {
      await connection.close();
      rethrow;
    } finally {
      await frames.cancel();
      await closed.cancel();
    }
    final resolvedRoute = registered.subdomain == null
        ? const RootRouteRequest()
        : SubdomainRouteRequest(registered.subdomain!);
    if (requestedRoute is AutoRouteRequest) {
      if (registered.subdomain == null) {
        await connection.close();
        throw ProxerError('Registered unexpected tunnel "root"');
      }
      requestedRoute = resolvedRoute;
    } else if (requestedRoute is RootRouteRequest &&
        registered.subdomain != null) {
      await connection.close();
      throw ProxerError(
        'Registered unexpected tunnel "${registered.subdomain}"',
      );
    } else if (requestedRoute case SubdomainRouteRequest(:final subdomain)) {
      if (registered.subdomain != subdomain) {
        await connection.close();
        throw ProxerError(
          'Registered unexpected tunnel "${registered.subdomain ?? "root"}"',
        );
      }
    }
    final forwarders = LocalForwarders(
      connection: connection,
      localPort: config.localPort,
      log: config.log,
      logError: config.logError,
    );
    config.log?.call('registered route=${_routeName(resolvedRoute)}');
    return _ActiveConnection(
      connection: connection,
      forwarders: forwarders,
      route: resolvedRoute,
    );
  }

  Future<void> cleanupActive() async {
    final current = active;
    active = null;
    if (current != null) {
      await current.forwarders.close();
    }
  }

  late void Function(_ActiveConnection value) watchClose;

  void scheduleReconnect() {
    if (closing || reconnectTimer != null) {
      return;
    }
    config.log?.call('reconnecting in ${config.reconnectDelayMs}ms');
    reconnectTimer = Timer(
      Duration(milliseconds: config.reconnectDelayMs),
      () async {
        reconnectTimer = null;
        if (closing) {
          return;
        }
        try {
          final next = await connect();
          if (closing) {
            await next.forwarders.close();
            await next.connection.close();
            return;
          }
          active = next;
          watchClose(next);
          config.log?.call('reconnected route=${_routeName(next.route)}');
        } on Object {
          scheduleReconnect();
        }
      },
    );
  }

  void onDisconnected(_ActiveConnection value) {
    if (!identical(active, value) || closing) {
      return;
    }
    unawaited(cleanupActive().then((_) => scheduleReconnect()));
  }

  watchClose = (_ActiveConnection value) {
    unawaited(value.connection.closed.first.then((_) => onDisconnected(value)));
  };

  active = await connect();
  watchClose(active!);
  final initialRoute = active!.route;

  return RunningTunnelClient(
    subdomain: initialRoute is SubdomainRouteRequest
        ? initialRoute.subdomain
        : null,
    close: () async {
      closing = true;
      reconnectTimer?.cancel();
      reconnectTimer = null;
      final current = active;
      await cleanupActive();
      await current?.connection.close();
    },
  );
}

String _routeName(HttpClientRouteRequest route) {
  return switch (route) {
    AutoRouteRequest() => 'auto',
    RootRouteRequest() => 'root',
    SubdomainRouteRequest(:final subdomain) => subdomain,
  };
}
